flash_attn/models/gpt.py

```diff
@@ -628,18 +628,24 @@ def forward(self, input_ids, position_ids=None, inference_params=None, last_toke
         last_token_only: whether to return the logit for the last token only,
             of shape (batch_size, vocab_size)
         """
+        assert input_ids.ndim == 2
+        b, slen = input_ids.shape
         hidden_states = self.transformer(
             input_ids, position_ids=position_ids, inference_params=inference_params
         )
-        if last_token_only:
-            hidden_states = hidden_states[:, -1]
         if self.project_out is not None:
             hidden_states = self.project_out(hidden_states)
+        # shape is either (b*slen/N, ft) in seqparallel
```

Comment:
nit: if `self.project_out` is not None and `last_token_only` is True, this structure might waste some compute for projecting the un-returned tokens. any chance we could move the hidden state reshape and extraction to before project_out?